framework/src/base/MooseApp.C

```diff
@@ -2083,112 +2089,101 @@ MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type,
+      if (!rm->attachGeometricEarly() && !attach_geometric_rm_final)
+      {
+        // Will attach them later (during algebraic). But also, we need to tell the mesh that we
+        // shouldn't be deleting remote elements yet
+        if (!mesh->getMeshPtr())
+          mooseError("We should have attached a MeshBase object to the mesh by now");
 
-          mesh->allowRemoteElementRemoval(false);
-        }
-        else
+        mesh->allowRemoteElementRemoval(false);
+      }
+      else
+      {
+        MeshBase & undisp_mesh_base = mesh->getMesh();
+        const DofMap * const undisp_nl_dof_map =
+            _executioner ? &_executioner->feProblem().systemBaseNonlinear().dofMap() : nullptr;
+        rm->init(undisp_mesh_base, undisp_nl_dof_map);
+        undisp_mesh_base.add_ghosting_functor(*rm);
+
+        // In the final stage, if there is a displaced mesh, we need to
+        // clone ghosting functors for displacedMesh
+        if (attach_geometric_rm_final && _action_warehouse.displacedMesh())
         {
-          MeshBase & mesh_base = mesh->getMesh();
-          rm->init(mesh_base);
-          mesh_base.add_ghosting_functor(*rm);
-
-          // In the final stage, if there is a displaced mesh, we need to
-          // clone ghosting functors for displacedMesh
-          if (attach_geometric_rm_final && _action_warehouse.displacedMesh())
-          {
-            std::shared_ptr<GhostingFunctor> clone_rm = rm->clone();
-            clone_rm->set_mesh(&_action_warehouse.displacedMesh()->getMesh());
-            _action_warehouse.displacedMesh()->getMesh().add_ghosting_functor(clone_rm);
-          }
-          else if (_action_warehouse.displacedMesh())
-            mooseError("Theh displaced mesh should not yet exist at the time that we are attaching "
-                       "geometric relationship managers.");
+          std::shared_ptr<GhostingFunctor> clone_gf = rm->clone();
+          auto clone_rm = std::static_pointer_cast<RelationshipManager>(clone_gf);
+          MeshBase & disp_mesh_base = _action_warehouse.displacedMesh()->getMesh();
+          const DofMap * disp_nl_dof_map = nullptr;
+          if (_executioner && _executioner->feProblem().getDisplacedProblem())
+            disp_nl_dof_map =
+                &_executioner->feProblem().getDisplacedProblem()->systemBaseNonlinear().dofMap();
+          clone_rm->init(disp_mesh_base, disp_nl_dof_map);
+          disp_mesh_base.add_ghosting_functor(clone_gf);
+          _undisp_to_disp_rms.emplace(rm.get(), clone_gf);
         }
+        else if (_action_warehouse.displacedMesh())
+          mooseError("The displaced mesh should not yet exist at the time that we are attaching "
+                     "early geometric relationship managers.");
       }
+    }
+    else // rm_type is algebraic or coupling
+    {
+      // Now we've built the problem, so we can use it
+      auto & problem = _executioner->feProblem();
```

Comment:
Should we check if _executioner exists? If it does not exist, we could error out